rfc/text/advanced/ap_10_rpc_progressive_calls.md

```diff
@@ -1,5 +1,230 @@
+A *Caller* may issue a progressive call. This can be useful in few cases:
 
-See this GitHub issue for more discussion: https://github.com/wamp-proto/wamp-proto/issues/167
+* Payload is too big to send it whole in one request.
+* Payload is long term-based, and it is required to start processing it earlier.
+* RPC is called too often. So overall processing can be optimized: there is no need to generate new id for request,
+  initiate structures for new call, etc.
+
+In this case, a procedure implemented by a *Callee* and registered at a *Dealer* may receive a progressive call.
+*Callee* can wait until it receives the whole payload and only after that process and sends result or 
+if *Callee*, *Caller* and *Dealer* supports `Progressive Call Results` feature it can even start sending results 
+back to *Caller* right after processing received payload chunks.
+This results in efficient two-way streams between *Caller* and *Callee*.
+
+**Feature Announcement**
+
+Support for this advanced feature MUST be announced by *Callers* (`role := "caller"`), *Callees* (`role := "callee"`) and *Dealers* (`role := "dealer"`) via
+
+{align="left"}
+HELLO.Details.roles.<role>.features.progressive_calls|bool := true
+
+**Message Flow**
+
+The message flow for progressive call when *Callee* waits for all chunks before processing:
+
+{align="left"}
+,------.            ,------.                 ,------.
+|Caller|            |Dealer|                 |Callee|
+`--+---'            `--+---'                 `--+---'
+   |  CALL (progress)  |                        |
+   | ----------------->|                        |
+   |                   | INVOCATION (progress)  |
+   |                   | ---------------------->|
+   |  CALL (progress)  |                        |
+   | ----------------->|                        |
+   |                   | INVOCATION (progress)  |
+   |                   | ---------------------->|
+   |                   |                        |
+   |       ...         |       ...              |
+   |                   |                        |
+   |  CALL (final)     |                        |
+   | ----------------->|                        |
+   |                   | INVOCATION (final)     |
+   |                   | ---------------------->|
+   |                   |                        |
+   |                   |  YIELD or ERROR        |
+   |                   | <----------------------|
+   |                   |                        |
+   |  RESULT or ERROR  |                        |
+   | <-----------------|                        |
+,--+---.            ,--+---.                 ,--+---.
+|Caller|            |Dealer|                 |Callee|
+`------'            `------'                 `------'
+
+As `progressive` call is still the same call *Callee* must send the same `Request|id` with every `CALL` and *Dealer*
+must also send the same `Request|id` with every `INVOCATION` to the *Callee*.
+
```

Comment:
Above diagram illustrates _one_ case for progressive calls. There are multiple more. There are these top level cases:

i) non-progressive call + non-progressive call results => a simple WAMP call
ii) non-progressive call + progressive call results
iii) progressive call + non-progressive call results => above message diagram
iv) progressive call + progressive call results => a bidirectionally streaming call

For iv), the crucial question we need to answer first is: what is the relation between individual CALLs (with progress) to individual INVOCATIONs, YIELDs and RESULTs. If any. The weakest restriction would be:

* the last CALL (final) must be forwarded in (a last) INVOCATION (final), and must be answered by RESULT (final), after which no more messages are valid for the (now completed) call

Sidenote: we need the progress flag on YIELD as well, but NOT on ERROR (there can be at most 1 ERROR for the whole call ... and ERROR ends the call)

Reply:
I have 3 and 4 cases schemes here. 1st is described in basic profile call and 2nd in progressive call results. So i think there is no need to copy paste them here.
And yep, we already have a `progress` flag in YIELD message, it's described also in progressive call results :)